commit/report.go

```diff
@@ -109,17 +278,30 @@ func (p *Plugin) Reports(
 	lggr.Debugw("transmission schedule override",
 		"transmissionSchedule", transmissionSchedule, "oracleIDToP2PID", p.oracleIDToP2PID)
 
-	lggr.Infow("commit plugin generated reports", "report", rep, "reportInfo", repInfo)
+	var reports []ocr3types.ReportPlus[[]byte]
 
-	return []ocr3types.ReportPlus[[]byte]{
-		{
-			ReportWithInfo: ocr3types.ReportWithInfo[[]byte]{
-				Report: encodedReport,
-				Info:   encodedInfo,
-			},
-			TransmissionScheduleOverride: transmissionSchedule,
-		},
-	}, nil
+	// The Solana report has some special limitations. Those limitations can be used if the offchain
+	maxRoots := p.offchainCfg.MaxMerkleRootsPerReport
+	if !p.offchainCfg.RMNEnabled && maxRoots != 0 {
+		if p.offchainCfg.MultipleReports {
+			lggr.Infow("building multiple reports")
+			reports, err = buildMultipleReports(ctx, lggr, p.reportCodec, transmissionSchedule, outcome, maxRoots)
+		} else {
+			lggr.Infow("building truncated reports")
+			reports, err = buildTruncatedReport(ctx, lggr, p.reportCodec, transmissionSchedule, outcome, maxRoots)
+		}
+	} else {
+		reports, err = buildStandardReport(ctx, lggr, p.reportCodec, transmissionSchedule, outcome)
+	}
```

Comment:
Can this kinda logic be in a helper?

Reply:
I added an abstraction and am now initializing this in the constructor. When we get to this point we'll simply call the handler function to build a report w/o needing to know which one we'll have.